csrc/device_lower/pass/alias_memory.cpp

```diff
@@ -1788,21 +1825,198 @@ class StackBasedSharedMemAllocator : kir::IrVisitor {
+      exprs_ = exprs;
+    } else {
+      if (isDebugDumpEnabled(DebugDumpOption::BufferReuseInfo)) {
+        debug() << "Ensuring syncs within these intervals:" << std::endl;
+        for (auto [last_read, first_write] : sync_intervals_) {
+          debug() << "  " << last_read << " - " << first_write << std::endl;
+        }
+      }
+      traverseAndInsert(exprs);
+    }
+  }
+
+  const std::unordered_set<Expr*>& insertedSyncs() const {
+    return inserted_syncs_;
+  }
+
+  const std::vector<Expr*>& modifiedExprs() const {
+    return exprs_;
+  }
+
+ private:
+  using kir::ExprMutator::dispatch;
+
+  void dispatch(Expr* expr) final {
+    // Skip inserted syncs
+    if (inserted_syncs_.find(expr) != inserted_syncs_.end()) {
+      if (isDebugDumpEnabled(DebugDumpOption::BufferReuseInfo)) {
+        debug() << "Skipping new sync expression " << expr->toString();
+      }
+      kir::ExprMutator::dispatch(expr);
+    }
+
+    position_ = allocation_info_map_.getScopeMap().getExprPos(expr);
+
+    // If this is an upcoming first write that has not yet been erased, it means
+    // we have not seen a sync in its interval. So we should insert a BlockSync
+    // before this expr.
+    if (upcoming_first_writes_.erase(position_)) {
+      if (isDebugDumpEnabled(DebugDumpOption::BufferReuseInfo)) {
+        debug() << "Inserting block sync before position " << position_
+                << std::endl;
+      }
+      auto new_sync = IrBuilder::create<kir::BlockSync>();
+      inserted_syncs_.insert(new_sync);
+      registerInsertBefore(expr, new_sync);
+    }
+
+    // If we have a sync at this location, we can clear any upcoming first
+    // writes since they can be considered safe.
+    if (lower_utils::hasBlockSync(expr, GpuLower::current()->threadPredMap())) {
+      if (isDebugDumpEnabled(DebugDumpOption::BufferReuseInfo)) {
+        debug() << "Found blocking expression at position " << position_
+                << std::endl;
+      }
+      upcoming_first_writes_.clear();
+    }
+
+    // If this is the lower endpoint of a sync interval, add the upper endpoint
+    auto range = sync_intervals_.equal_range(position_);
+    for (auto& it = range.first; it != range.second; ++it) {
+      if (isDebugDumpEnabled(DebugDumpOption::BufferReuseInfo)) {
+        debug() << "Found dependency last read at position " << position_
+                << " corresponding to first write at " << it->second
+                << std::endl;
+      }
+      upcoming_first_writes_.insert(it->second);
+    }
+
+    kir::ExprMutator::dispatch(expr);
+  }
+
+ private:
+  const AllocationInfoMap& allocation_info_map_;
+
+  // This holds intervals in which we need to ensure a sync exists. All
+  // threads in a block should arrive at the start of each interval before any
+  // thread proceeds past the end of the interval.
+  std::unordered_multimap<int, int> sync_intervals_;
+
+  // Position within the traversal
+  int position_ = -1;
```

Comment:
Does this need to be a member?

Reply:
It does not. This was leftover from a refactor. Fixing..

Reply:
Done